services/publishing/Tweek.Publishing.Service/Helpers/HooksHelper.cs

```diff
@@ -0,0 +1,153 @@
+    private static readonly MetricTags _metricsSuccess = new MetricTags("Status", "Success");
+    private static readonly MetricTags _metricsFailure = new MetricTags("Status", "Failure");
+
+    private HooksHelper(Func<string, Task<string>> gitExecutor, IMetrics metrics, ILogger logger) {
+      this._logger = logger;
+      this._git = gitExecutor;
+      this._metrics = metrics;
+      this._keysRegex = new Regex(@"(?:implementations/jpad/|manifests/)(.*)\..*", RegexOptions.Compiled);
+    }
+
+    public static void Initialize(Func<string, Task<string>> gitExecutor, IMetrics metrics, ILogger logger = null) {
+      _instance = new HooksHelper(gitExecutor, metrics, logger ?? NullLogger.Instance);
+    }
+
+    public static HooksHelper GetInstance() {
+      return _instance;
+    }
+
+    public async Task TriggerNotificationHooksForCommit(string commitId) {
+      try {
+        var keyPaths = await _GetKeyPathsFromCommit(commitId);
+        var allHooks = await _GetAllHooks(commitId);
+
+        var keyPathsByHook = _AggregateKeyPathsByHook(keyPaths, allHooks);
+        var usedKeyPaths = _GetUsedKeyPaths(keyPathsByHook);
+        var keyPathsData = await _GetKeyPathsData(usedKeyPaths, commitId);
+
+        var hooksWithData = _GetHooksWithKeyPathData(keyPathsByHook, keyPathsData);
+        await _TriggerHooks(hooksWithData, commitId);
+      } catch (Exception ex) {
+        _logger.LogError(ex, $"Failed triggering notification hooks for commit {commitId}");
+        _metrics.Measure.Counter.Increment(_hooksMetric, _metricsFailure);
+      }
+    }
+
+    private async Task _TriggerHooks(Dictionary<Hook, string> hooksWithData, string commitId) {
+      var triggerTasks = hooksWithData.Select( kvp => kvp.Key.Trigger(kvp.Value) );
+
+      foreach (var triggerTask in triggerTasks) {
+        try {
+          await triggerTask;
+
+          _metrics.Measure.Counter.Increment(_hooksMetric, _metricsSuccess);
+        } catch (Exception ex) {
+          _logger.LogError(ex, $"Failed triggering a notification hook for commit {commitId}");
+          _metrics.Measure.Counter.Increment(_hooksMetric, _metricsFailure);
+        }
+      }
+    }
+
+    private Dictionary<Hook, string> _GetHooksWithKeyPathData(
+      Dictionary< Hook, HashSet<string> > keyPathsByHook,
+      Dictionary<string, KeyPathData> keyPathsData
+    ) {
+      return keyPathsByHook.Aggregate(new Dictionary<Hook, string>(), (acc, kvp) => {
+        var hookData = kvp.Value.Select( keyPath => keyPathsData[keyPath] );
+        var hookDataJson = JsonConvert.SerializeObject(hookData);
+
+        acc.Add(kvp.Key, hookDataJson);
+        return acc;
+      });
+    }
+
+    private async Task< Dictionary<string, KeyPathData> > _GetKeyPathsData(IEnumerable<string> keyPaths, string commitId) {
+      var keyPathsDataDict = new Dictionary<string, KeyPathData>(keyPaths.Count());
+
+      var keyDataTasks = keyPaths.Select(keyPath => new Task<string>[] {
```

Comment:
.Net has support for named tuples, so it can be clear to understand compare to array.
(keyPath: Task.FromResult(keyPath), implementation : _git($"show {commitId}:implementations/jpad/{keyPath}.jpad"), 
_git($"show {commitId}:manifests/{keyPath}.json")
 );

And then it can be (await  tasks.keyPath, await  tasks.implementation, await  tasks.manifest)

Reply:
Nice, didn't know about named tuples, will be very useful